File: objectivec/pack/LongArrayPool.m


#import <Foundation/Foundation.h>
#import <os/lock.h>
#import "LongArrayPool.h"


static const int DEFAULT_CAPACITY = 16;
static const int SECOND_SIZE = 64;
static const int SECOND_CAPACITY = 4;

static int defaultCount = 0;
static uint64_t* defaultArrays[DEFAULT_CAPACITY];

static int secondCount = 0;
static uint64_t* secondArrays[SECOND_CAPACITY];

static os_unfair_lock lock1;
static os_unfair_lock lock2;

@implementation LongArrayPool

+(uint64_t*)getLongArray:(int)size{
    if(size <= LONG_ARRAY_DEFAULT_SIZE){
        uint64_t* a;
        os_unfair_lock_lock(&lock1);
        if(defaultCount > 0){
            defaultCount--;
            a = defaultArrays[defaultCount];
        }else{
            a = (uint64_t*)malloc(32<<3);
        }
        os_unfair_lock_unlock(&lock1);
        return a;
    }
    if(size <= SECOND_SIZE){
        uint64_t* a;
        os_unfair_lock_lock(&lock2);
        if(secondCount > 0){
            secondCount--;
            a = secondArrays[secondCount];
        }else{
            a = (uint64_t*)malloc(64<<3);
        }
        os_unfair_lock_unlock(&lock2);
        return a;
    }
    if(size <= 128){
        return (uint64_t*)malloc(128<<3);
    }
    return (uint64_t*)malloc(256<<3);
}

+(void)recycleLongArray:(uint64_t*)a withLen:(int)len{
    if(a == NULL){
        return;
    }
    if(len == LONG_ARRAY_DEFAULT_SIZE){
        os_unfair_lock_lock(&lock1);
        if(defaultCount < DEFAULT_CAPACITY){
            defaultArrays[defaultCount] =a;
            defaultCount++;
        }else{
            free(a);
        }
        os_unfair_lock_unlock(&lock1);
    }else if(len == SECOND_SIZE){
        os_unfair_lock_lock(&lock2);
        if(secondCount < SECOND_CAPACITY){
            secondArrays[secondCount] =a;
            secondCount++;
        }else{
            free(a);
        }
        os_unfair_lock_unlock(&lock2);
    }else{
        free(a);
    }
}

@end

File: go/pack/LongArrayPool.go
package pack

import "sync"

const (
	default_size     int32 = 32
	default_capacity int32 = 24
	second_size      int32 = 64
	second_capacity  int32 = 8
)

var defaultCount int32 = 0
var defaultArrays [default_capacity][]uint64

var secondCount int32 = 0
var secondArrays [second_capacity][]uint64

var lMutex1 sync.Mutex
var lMutex2 sync.Mutex

func getLongArray(size int32) []uint64 {
	if size <= default_size {
		var a []uint64
		lMutex1.Lock()
		if defaultCount > 0 {
			defaultCount--
			a = defaultArrays[defaultCount]
			defaultArrays[defaultCount] = nil

		} else {
			a = make([]uint64, default_size)
		}
		lMutex1.Unlock()
		return a
	}
	if size <= second_size {
		var a []uint64
		lMutex2.Lock()
		if secondCount > 0 {
			secondCount--
			a = secondArrays[secondCount]
			secondArrays[secondCount] = nil
		} else {
			a = make([]uint64, second_size)
		}
		lMutex2.Unlock()
		return a
	}
	if size <= 128 {
		return make([]uint64, 128)
	}
	return make([]uint64, 256)
}

func recycleLongArray(a []uint64) {
	if a == nil {
		return
	}
	size := int32(len(a))

	if size == default_size {
		lMutex1.Lock()
		if defaultCount < default_capacity {
			defaultArrays[defaultCount] = a
			defaultCount++
		}
		lMutex1.Unlock()
	} else if size == second_size {
		lMutex2.Lock()
		if secondCount < second_capacity {
			secondArrays[secondCount] = a
			secondCount++
		}
		lMutex2.Unlock()
	}
}
